doc/man/builddbm.8
.\" $Id$
.ds RD /usr/local/etc/raddb
.ds RL /usr/adm
.ds RP /usr/local/etc/raddb
.TH BUILDDBM 8 "March 10, 2000" "FSF"
.SH NAME
builddbm - build a DBM version of Radius users database
.SH SYNOPSIS
\fBbuilddbm\fR
[\fB-d\fR config_dir]
[input_file]
.PP
\fBbuilddbm\fR
\fB-h\fR
.SH DESCRIPTION
\fBBuilddbm\fR converts the plaintext \fBRadius\fR users database into
DBM files. Some versions of \fBRadius\fR daemon have used this to
speed up the access to the users database.
.PP
Howewer in this implementation of \fBRadius\fR daemon things go the
other way around. The server reads entire plaintext database, converts
it into internal form and stores into hash table which provides for
fast access. Actually, using DBM version of the users database slows
down the access unless the machine which runs \fBRadius\fR daemon is
short of address space for the daemon to store the users database into.
.PP
Optionally, \fBRadius\fR daemon can use both plaintext and DBM
databases, in which case it first looks up the user in the former, and
if not found, continues search in the latter.
.SH OPERATION
When used without arguments, \fBbuilddbm\fR utility attempts to
convert file \fB\*(RD/users\fR into \fB\*(RD/users.db\fR or
\fB\*(RD/users.pag, \*(RD/users.dir\fR pair depending on the
version of DBM library used.
.PP
If used with one argument, the argument is taken as the name of the
plaintext database file to operate upon.
.SH OPTIONS
.IP -d
Specifies alternate directory for the \fBRadius\fR users database
files. This defaults to \*(RD.
.IP -i
Specifies alternate input file.
.IP -h
Outputs short usage summary and exits with 0 exit code.
.SH FILES
.IP \*(RD/users
Default plaintext database.
.IP \*(RD/users.db
Default output file.
.SH BUGS
The information in this manpage may be obsolete or incomplete. Please
refer to texinfo documentation for full information about GNU Radius
tools.
.SH AUTHORS
Sergey Poznyakoff, gray@farlep.net

.SH SEE ALSO
users(5rad), radiusd(8rad), config(5rad)




NEWS
GNU Radius NEWS -- history of user-visible changes. 2004-04-23
Copyright (C) 2002,2003,2004 Free Software Foundation, Inc.
See the end of file for copying conditions.

Please send radius bug reports to <bug-gnu-radius@gnu.org>
^L
Version 1.2.3:

* libgnuradius

This is a library of functions for creation, handling and sending
requests using RADIUS protocol.

All programs have been rewritten to link with libgnuradius. On most
sites this will mean linking against a shared library, which will
reduce the size of the executables.

* gnuradius.scm

This is a guile module allowing to use libgnuradius functions. It
supercedes radscm program, which has been removed.

* Rewrite language

** Added i18n support

** New built-in functions:

*** Functions to access internal fields of a RADIUS request.
*** Interfaces to the Radius NAS database (raddb/naslist).
*** Interfaces to DNS lookup functions.

* radiusd

** Improved checking for multiple logins. Previous versions relied entirely
on the contents of /var/log/radutmp file. Starting at this version,
radiusd offers at least two methods of checking for multiple logins:
using the traditional radutmp file and using the SQL database. New
keywords has been added to the sqlserver file that declare the SQL
queries to be used when retrieving the information about currently
active sessions.

More methods of checking will be added in future versions.

** When an unsupported authentication type is requested, radiusd first
checks if an extension Scheme module is provided that handles that
authentication type. If such module is found, it is invoked to handle
the authentication.

** System accounting can be turned off by specifying `system no;' in
`acct' section of raddb/config. 

** New configuration statement 'load-module' allows to load arbitrary
Scheme modules.

** Detailed log filenames are configurable via `detail-file-name'
statements in `auth' and `acct' sections of raddb/config.

** Support for Guile versions prior to 1.6 has been withdrawn.

* Bugfixes

** Allow to omit port numbers in `listen' statements (raddb/config), as
described in the documentation.
    


Version 1.2:

* Important notes for upgrading from earlier versions of GNU Radius

** Upgrade your dictionaries.

This version does a very strict semantic checking on the dictionary
files. Old dictionaries may produce lots of warnings and
errors, therefore you are advised to upgrade them. To do so,
run `make update-dictionary'. It will update your main 
(raddb/dictionary) and secondary dictionaries (raddb/dict/*),
preserving any changes you did to the former. 

Notice, that, starting from this version, the default main dictionary
file will not contain any attribute declarations, instead it will
merely include the necessary dictionaries. However, you may add your
declarations to it and modify it to your linking.

Any modifications to the dictionary files in raddb/dict are now
strongly discouraged, as these files will be updated by each new
release. Put your modifications to the main dictionary file only.
This file will remain unchanged during subsequent installs.

Two new statements are provided to help you customize your dictionary.
See section `** Dictionary' below.

** Update your raddb/config and raddb/rewrite

The rewrite procedures are no longer stored in a single file
`radiusd/rewrite', instead they may occupy several source file.
The rewrite procedures shipped with this version of GNU Radius
are installed in ${prefix}/share/radius/1.2/rewrite. We strongly
advise you to add the following statement to your raddb/config:

rewrite {
        load "checknas.rw";
        load "log-hook.rw";
};

You may then remove the old raddb/rewrite file, *unless* you have
modified it.

If you have modified it, follow the procedure described below:

1. Load this file, by adding the following statement to the above
`rewrite' block:

	load "rewrite";

2. Make sure to remove from this file or rename the following functions:
check_snmp_u, check_snmp_s, decimal, check_max_finger, check_unix.

** Tailor log file format.

The format of log file is now fully customizable via a special Rewrite
procedures, called "hooks" (search for logging-hook in the text below
for the detailed description). To revert to logfile format used in
versions 1.0-1.1, add the following two statements to your raddb/config,
right after `logging {' line:

       	prefix-hook "compat_log_prefix";
	suffix-hook "compat_log_suffix";

** Using Guile extensions

If you are using Guile extensions, notice that the default location
for Radius scheme modules is now ${prefix}/share/radius/1.2/scheme.
	
* New features:

** Support for Status-Server request.

** Support for Tunnel Protocol (RFC 2865)

** The main configuration file raddb/config allows to specify several
Rewrite source files to be loaded instead of the single predefined
file raddb/rewrite.

** Arbitrary Rewrite expressions may be used in A/V pairs, for example:

DEFAULT	     Auth-Type = SQL
	     Framed-IP-Address =
	           "=10.10.10.1 + %[NAS-Port-Id]/16 + %[NAS-Port-Id]"

** Test shell mode considerably improved.

** The internal attributes do not appear in detailed logs. The
special flag 'l' has been added to the dictionary syntax to
turn this feature off. Thus, for example:

ATTRIBUTE Orig-User-Name  2101	string	- [L--R-R]l

will make Orig-User-Name attribute visible in detailed log files.

** Changes to Rewrite language

*** usual dotted-quad notation may be used to represent IP numbers.

*** a number of new built-in functions is added

*** #pragma regexp allows to alter the type of regular expressions
used. The rest of line following `#pragma regexp' must consist of
whitespace-separated regexp modifiers. Each regexp modifier is one
of the following keywords:

  extended      Use extended regular expressions
  icase         Ignore case
  newline       Match-any-character operators don't match a newline.

optionally preceded by plus, indicating enabling the feature, or

builddbm/builddbm.c
/* This file is part of GNU RADIUS.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 */
/*      RADIUS builddbm utility */

#ifndef lint
static char rcsid[] =
"@(#) $Id$";
#endif

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <netinet/in.h>

#include <stdio.h>
#include <netdb.h>
#include <strings.h>
#include <pwd.h>
#include <time.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>

#include <radiusd.h>
#include <raddbm.h>
#include <parser.h>

char		*progname;
int		debug_flag;
char		*radius_dir = RADIUS_DIR;
#ifdef DBM
char            *pag_file;
char            *dir_file;
#endif
char            *db_file;
char            *users_file;
char            *input="users";

typedef struct {
	DBM_FILE dbmfile;
	int begno;   /* ordinal number of next BEGIN entry */
	int defno;   /* ordinal number of next DEFAULT entry */
} DBM_closure;

int pair_buffer[RAD_BUFFER_SIZE];
void usage();
int add_user(DBM_closure *closure, int line,
	     char *name, VALUE_PAIR *check, VALUE_PAIR *reply);

int
main(argc, argv)
	int argc;
	char **argv;
{
	int     c;
	DBM_closure closure;
#ifdef DBM
	int fd;
#endif
	
	initlog(argv[0]);

	while ((c = getopt(argc, argv, "hir:")) != EOF) {
		switch (c) {
		case 'h':
			usage();
			/*NOTREACHED*/
		case 'i':
			users_file = optarg;
			break;
		case 'r':
			radius_dir = optarg;
			break;
		default:
			return -1;
		}
	}

	radpath_init();

	if (!users_file && argv[optind])
		users_file = argv[optind];

	if (dict_init()) {
		radlog(L_ERR, _("error reading dictionary file"));
		return 1;
	}
	
	/* Make filenames: */
	if (!users_file)
		users_file = mkfilename(radius_dir, input);

	db_file = mkfilename(radius_dir, "users");

	/*
	 *	Initialize a new, empty database.
	 */
	closure.defno = closure.begno = 0;
	if (create_dbm(db_file, &closure.dbmfile)) {
		radlog(L_ERR|L_PERROR, _("can't open `%s'"), db_file);
		return 1;
	}

	parse_file(users_file, &closure, add_user);
	
	close_dbm(closure.dbmfile);

	return 0;
}

#define NINT(n) ((n) + sizeof(int) - 1)/sizeof(int)

/*ARGSUSED*/
int
add_user(closure, line, name, check, reply)
	DBM_closure *closure;
	int line;
	char *name;
	VALUE_PAIR *check, *reply;
{
	int     check_len;
	int     reply_len;
	VALUE_PAIR *vp;
	int     *q;
	datum	named;
	datum	contentd;
	
	check_len = list_length(check);
	reply_len = list_length(reply);

	if (2 + check_len + reply_len > sizeof(pair_buffer)) {
		radlog(L_ERR, _("%s:%d: too many attributes"),
		       source_filename, source_line_num);
		return -1;
	}

	q = pair_buffer;
	*q++ = check_len;
	for (vp = check; vp; vp = vp->next) {
		*q++ = vp->attribute;
		*q++ = vp->type;
		*q++ = vp->operator;
		if (vp->type == PW_TYPE_STRING) {
			strcpy((char*)q, vp->strvalue);
			q += NINT(vp->strlength+1);
		} else
			*q++ = vp->lvalue;
	}
	*q++ = reply_len;
	for (vp = reply; vp; vp = vp->next) {
		*q++ = vp->attribute;
		*q++ = vp->type;
		*q++ = vp->operator;
		if (vp->type == PW_TYPE_STRING) {
			strcpy((char*)q, vp->strvalue);
			q += NINT(vp->strlength+1);
		} else
			*q++ = vp->lvalue;
	}
	
	if (strncmp(name, "DEFAULT", 7) == 0) 
		sprintf(name, "DEFAULT%d", closure->defno++);
	else if (strncmp(name, "BEGIN", 5) == 0) 
		sprintf(name, "BEGIN%d", closure->begno++);
	
	named.dptr = name;
	named.dsize = strlen(name);
	contentd.dptr = (char*)pair_buffer;
	contentd.dsize = (2 + check_len + reply_len) * sizeof(int);
	if (insert_dbm(closure->dbmfile, named, contentd)) {
		radlog(L_ERR, _("can't store datum for %s"), name);
		exit(1);
	}
	return 0;
}



/*
 *	Fixup a check line.
 *	If Password or Crypt-Password is set, but there is no
 *	Auth-Type, add one (kludge!).
 */
void
auth_type_fixup(check)
	VALUE_PAIR *check;
{
	VALUE_PAIR	*vp;
	VALUE_PAIR	*c = NULL;
	int		n;

	/*
	 *	See if a password is present. Return right away
	 *	if we see Auth-Type.
	 */
	for (vp = check; vp; vp = vp->next) {
		if (vp->attribute == DA_AUTH_TYPE)
			return;
		if (vp->attribute == DA_PASSWORD) {
			c = vp;
			n = DV_AUTH_TYPE_LOCAL;
		}
		if (vp->attribute == DA_CRYPT_PASSWORD) {
			c = vp;
			n = DV_AUTH_TYPE_CRYPT_LOCAL;
		}
	}

	if (c == NULL)
		return;

	/*
	 *	Add an Auth-Type attribute.
	 */
	vp = alloc_pair();
	memset(vp, 0, sizeof(VALUE_PAIR));
	strcpy(vp->name, "Auth-Type");
	vp->attribute = DA_AUTH_TYPE;
	vp->type = PW_TYPE_INTEGER;
	vp->lvalue = n;

	vp->next = c;
	c = vp->next;

}

int
list_length(vp)
	VALUE_PAIR *vp;
{
	int len;
	
	for (len = 0; vp; vp = vp->next) {
		len += 3;
		if (vp->type == PW_TYPE_STRING)
			len += NINT(vp->strlength + 1);
		else
			len++;
	}
	return len;
}

/* needed by users.y */
int
master_process()
{	
	return 1;
}

void
usage()
{
	printf(_("usage: builddbm [-i input_file][-r radius_dir]\n"));
}
